pkg/admission/cert/writer/secret.go

```diff
@@ -0,0 +1,246 @@
+func (s *SecretCertWriter) EnsureCerts(webhookConfig runtime.Object) error {
+	if webhookConfig == nil {
+		return errors.New("unexpected nil webhook configuration object")
+	}
+
+	secretWebhookMap := map[string]*webhookAndSecret{}
+	accessor, err := meta.Accessor(webhookConfig)
+	if err != nil {
+		return err
+	}
+	annotations := accessor.GetAnnotations()
+	// Parse the annotations to extract info
+	s.parseAnnotations(annotations, secretWebhookMap)
+
+	webhooks, err := getWebhooksFromObject(webhookConfig)
+	if err != nil {
+		return err
+	}
+	for i, webhook := range webhooks {
+		if s, found := secretWebhookMap[webhook.Name]; found {
+			s.webhook = &webhooks[i]
+		}
+	}
+
+	// validation
+	for k, v := range secretWebhookMap {
+		if v.webhook == nil {
+			return fmt.Errorf("expecting a webhook named %q", k)
+		}
+	}
+
+	certGenerator := s.CertGenerator
+	if s.CertGenerator == nil {
+		certGenerator = &generator.SelfSignedCertGenerator{}
+	}
+
+	srw := &secretReadWriter{
+		client:        s.Client,
+		certGenerator: certGenerator,
+		webhookConfig: webhookConfig,
+		webhookMap:    secretWebhookMap,
+	}
+	return srw.ensureCert()
+}
+
+func (s *SecretCertWriter) parseAnnotations(annotations map[string]string, secretWebhookMap map[string]*webhookAndSecret) {
+	for k, v := range annotations {
+		if strings.HasPrefix(k, SecretCertProvisionAnnotationKeyPrefix) {
+			webhookName := strings.TrimPrefix(k, SecretCertProvisionAnnotationKeyPrefix)
+			secretWebhookMap[webhookName] = &webhookAndSecret{
+				secret: types.NewNamespacedNameFromString(v),
+			}
+		}
+	}
+}
+
+func (s *secretReadWriter) ensureCert() error {
+	for _, v := range s.webhookMap {
+		err := handleCommon(v.webhook, s)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// secretReadWriter deals with writing to the k8s secrets.
+type secretReadWriter struct {
+	client        client.Client
+	certGenerator generator.CertGenerator
+
+	webhookConfig runtime.Object
+	webhookMap    map[string]*webhookAndSecret
+}
+
+type webhookAndSecret struct {
+	webhook *admissionregistrationv1beta1.Webhook
+	secret  apitypes.NamespacedName
+}
+
+var _ certReadWriter = &secretReadWriter{}
+
+func (s *secretReadWriter) write(webhookName string) (
+	*generator.Artifacts, error) {
+	v := s.webhookMap[webhookName]
+
+	webhook := v.webhook
+	commonName, err := dnsNameForWebhook(&webhook.ClientConfig)
+	if err != nil {
+		return nil, err
+	}
+	certs, err := s.certGenerator.Generate(commonName)
+	if err != nil {
+		return nil, err
+	}
+	secret := certsToSecret(certs, v.secret)
+	// TODO: fix it: see the TODO in the method
+	err = setOwnerRef(secret, s.webhookConfig)
+	if err != nil {
+		return nil, err
+	}
+	err = s.client.Create(nil, secret)
+	return certs, err
+}
+
+func (s *secretReadWriter) overwrite(webhookName string) (
+	*generator.Artifacts, error) {
+	v := s.webhookMap[webhookName]
+
+	webhook := v.webhook
```

Comment:
Most of this code is duplicated in write().  Please pull it out into a function.